File: backend/app/schemas/comment.py
from datetime import datetime
from typing import List, Optional
from uuid import UUID
from pydantic import BaseModel, Field
from app.schemas.user import UserBrief


class SkillCommentCreateRequest(BaseModel):
    content: str = Field(min_length=1, max_length=5000)
    parent_id: Optional[UUID] = None


class SkillCommentResponse(BaseModel):
    id: UUID
    skill_id: UUID
    parent_id: Optional[UUID] = None
    content: str
    author: UserBrief
    children: List["SkillCommentResponse"] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


SkillCommentResponse.model_rebuild()

File: packages/contract/src/comment.ts
import { z } from 'zod';
import { userBriefSchema, type UserBrief } from './user.js';

export interface SkillComment {
  id: string;
  skill_id: string;
  parent_id: string | null;
  content: string;
  author: UserBrief;
  children: SkillComment[];
  created_at: string;
  updated_at: string;
}

export const skillCommentSchema: z.ZodType<SkillComment> = z.lazy(() =>
  z.object({
    id: z.string().uuid(),
    skill_id: z.string().uuid(),
    parent_id: z.string().uuid().nullable(),
    content: z.string(),
    author: userBriefSchema,
    children: z.array(skillCommentSchema),
    created_at: z.string(),
    updated_at: z.string(),
  }),
);

export const createCommentInputSchema = z.object({
  content: z.string().min(1).max(2000),
  parent_id: z.string().uuid().optional(),
});
export type CreateCommentInput = z.infer<typeof createCommentInputSchema>;
